torch/nn/utils/attention.py

```diff
@@ -0,0 +1,253 @@
+    """A bias that is a tensor"""
+
+    def __init__(self, bias: torch.Tensor):
+        self.bias = bias
+
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        return self.bias
+
+    def needs_materialization(self) -> bool:
+        return True
+
+    @staticmethod
+    def dispatch(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_mask: "AttnBias",
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError(
+            "TensorBias requires materialization, so this should never be called!"
+        )
+
+    def __repr__(self) -> str:
+        return f"TensorBias(bias={self.bias})"
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func != torch.nn.functional.scaled_dot_product_attention:
+            return NotImplemented
+        args = pytree.tree_map_only(
+            TensorBias, lambda x: _materialize_if_needed(x), args
+        )
+        kwargs = pytree.tree_map_only(
+            TensorBias, lambda x: _materialize_if_needed(x), kwargs
+        )
+        return func(*args, **kwargs)
+
+
+class CausalVariant(IntEnum):
+    """Enum for causal variants"""
+
+    UPPER_LEFT = 1
+    LOWER_RIGHT = 2
+
+
+class CausalBias(AttnBias):
+    """A bias representing causal attention patterns"""
+
+    def __init__(self, variant: CausalVariant, seq_len_q: int, seq_len_kv: int):
+        assert isinstance(variant, CausalVariant)
+        self.variant = variant
+        self.seq_len_q = seq_len_q
+        self.seq_len_kv = seq_len_kv
+        if seq_len_q > seq_len_kv and variant == CausalVariant.LOWER_RIGHT:
+            warn(
+                "Lower right causal bias will produce NaNs in the output when seq_len_q > seq_len_kv!"
+            )
+
+    def _upper_left(self, device: torch.device) -> torch.Tensor:
+        """Upper left causal bias"""
+        return torch.tril(
+            torch.ones(self.seq_len_q, self.seq_len_kv, device=device, dtype=torch.bool)
+        )
+
+    def _lower_right(self, device: torch.device) -> torch.Tensor:
+        """Lower right causal bias"""
+        diagonal_offset = self.seq_len_kv - self.seq_len_q
+        return torch.tril(
+            torch.ones(
+                self.seq_len_q, self.seq_len_kv, device=device, dtype=torch.bool
+            ),
+            diagonal=diagonal_offset,
+        )
+
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        if device is None:
+            device = torch.device("cpu")
+        if self.variant == CausalVariant.UPPER_LEFT:
+            return self._upper_left(device)
+        elif self.variant == CausalVariant.LOWER_RIGHT:
+            return self._lower_right(device)
+
+    def needs_materialization(self) -> bool:
+        return False
```

Comment:
This makes it seem to me like `materialize()` never needs to be called, which can't be the case since it has an implementation. maybe all the more reason to streamline `needs_materialization()` and `materialize()` into a single API

Reply:
My thinking here was, although for the standard path of SDPA you never need to materialize the bias, there may be cases where users want to interact with an instance outside of SDPA. However, at least for my implementation I made torch_function disable everything but SDPA this makes less sense..